homework2/task01/project1/task1.py

```diff
@@ -0,0 +1,90 @@
+"""
+Given a file containing text. Complete using only default collections.
+
+1) Find 10 longest words consisting from largest amount of unique symbols
+2) Find rarest symbol for document
+3) Count every punctuation char
+4) Count every non ascii char
+5) Find most common non ascii char for document
+"""
+import string
+from typing import List
+
+
+def file_presettings(file_path: str) -> str:
+    """Return unpacked file, encoded in readable format without punctuation."""
+    with open(file_path, "r", encoding="unicode_escape") as f:
+        text = f.read()
+        for p in string.punctuation:
+            if p in text:
+                text = text.replace(p, "")
+        return text
+
+
+def get_unique_word_symbols(i: str) -> (int):
+    """Return amount of unique symbols."""
+    return len("".join(set(i)))
+
+
+def get_longest_diverse_words(file_path: str) -> List[str]:
+    """Return 10 longest diverse words."""
+    text = file_presettings(file_path)
+    all_words = list(dict.fromkeys(text.split()))
+    all_words.sort(key=get_unique_word_symbols, reverse=True)
+    return all_words[:10]
+
+
+def get_rarest_char(file_path: str) -> str:
+    """Return one single rarest char."""
+    with open(file_path, "r", encoding="unicode_escape") as f:
+        text = f.read()
+        all_chars = {}
+        for i in text:
+            if i not in all_chars:
+                all_chars[i] = 1
+            else:
+                all_chars[i] += 1
+        min_value = min(all_chars.values())
+        key_list = list(all_chars.keys())
+        val_list = list(all_chars.values())
+        return key_list[val_list.index(min_value)]
```

Comment:
It is still `O(n) + O(n)`.  It could be done in a single iteration.

Reply:
I changed now, please, see. I am out of ideas how else in a single iteration.